File: Camp Progress/207. Course Schedule.cpp
// Question Link: https://leetcode.com/problems/course-schedule

/*
    We need to detect if given graph is DAG (Directed Acyclic Graph)
    To detect that, we need to check if graph has any cycle
*/
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        // Build Directed Acyclic Graph
        vector<vector<int>> DAG(numCourses);
        for (int i = 0; i < prerequisites.size(); i++) {
            int from = prerequisites[i][1];
            int to = prerequisites[i][0];
            DAG[from].push_back( to );
        }
        
        vector<bool> seen(numCourses, false);
        for (int ithNode = 0; ithNode < numCourses; ithNode++) {
            if (seen[ithNode] == false) {
                vector<bool> seenForCycle(numCourses, false);
                bool hasCycle = dfs(ithNode, DAG, seen, seenForCycle);
                if (hasCycle) {
                    return false;
                }
            }
        }
        
        return true;
    }
    
    bool dfs(int cur, vector<vector<int>> &DAG, vector<bool> &seen, vector<bool> &seenForCycle) {
        
        seen[cur] = true;
        seenForCycle[cur] = true;
        
        bool hasCycle = false;
        for (int &neighbor : DAG[cur]) {
            if (seen[neighbor] == false) {
                hasCycle = dfs(neighbor, DAG, seen, seenForCycle);
                if (hasCycle) {
                    break;
                }
            }
            else if (seenForCycle[neighbor] == true) {
                hasCycle = true;
                break;
            }
        }
        
        // Backtracking
        seenForCycle[cur] = false;
        
        return hasCycle;
    }
};

File: Python3 Solutions/207. Course Schedule.py
# Question Link: https://leetcode.com/problems/course-schedule

"""
    We need to detect if given graph is DAG (Directed Acyclic Graph)
    To detect that, we need to check if graph has any cycle
"""

# BFS Solution
from collections import deque

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = [[] for _ in range(numCourses)]
        indegrees = [0] * numCourses
        for dest, src in prerequisites:
            graph[src].append(dest)
            indegrees[dest] += 1
        
        todo = deque()
        for i in range(numCourses):
            if indegrees[i] == 0:
                todo.append(i)
                
        nodesProcesses = 0
        while todo:
            cur = todo.pop()
            nodesProcesses += 1
            for neighbor in graph[cur]:
                indegrees[neighbor] -= 1
                if indegrees[neighbor] == 0:
                    todo.append(neighbor)
                
        return nodesProcesses == numCourses

    
# DFS Solution
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        
        dag = [[] for x in range(numCourses)]
        for dest, source in prerequisites:
            dag[dest].append(source)
            
        seen = [False] * numCourses
        cycleSeen = [False] * numCourses
            
        for i in range(0, numCourses):
            if seen[i] == False:
                hasCycle = self.dfs(i, dag, seen, cycleSeen)
                if hasCycle:
                    return False
                
        return True
    
    def dfs(self, cur, dag, seen, cycleSeen) -> bool:
        seen[cur] = True
        cycleSeen[cur] = True
        
        for neigh in dag[cur]:
            if seen[neigh] == False:
                hasCycle = self.dfs(neigh, dag, seen, cycleSeen)
                if hasCycle:
                    return True
                
            elif cycleSeen[neigh] == True:
                return True
        
        cycleSeen[cur] = False
        return False
